File: app/src/test/java/com/soundcloud/android/discovery/DiscoveryAdapterTest.java
package com.soundcloud.android.discovery;


import static java.util.Arrays.asList;
import static org.assertj.core.api.Java6Assertions.assertThat;

import com.soundcloud.android.search.SearchItemRenderer;
import com.soundcloud.android.testsupport.AndroidUnitTest;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;

public class DiscoveryAdapterTest extends AndroidUnitTest {

    @Mock SearchItemRenderer<DiscoveryCardViewModel> searchItemRenderer;
    @Mock SingleSelectionContentCardRenderer singleSelectionContentCardRenderer;
    @Mock MultipleContentSelectionCardRenderer multipleContentSelectionCardRenderer;
    @Mock EmptyCardRenderer emptyCardRenderer;
    @Mock SearchItemRenderer.SearchListener searchListener;

    final DiscoveryCardViewModel.MultipleContentSelectionCard multipleContentSelectionCard = DiscoveryFixtures.multiContentSelectionCardViewModel();
    final DiscoveryCardViewModel.SingleContentSelectionCard singleContentSelectionCard = DiscoveryFixtures.singleContentSelectionCardViewModel();

    private DiscoveryAdapter adapter;
    private DiscoveryCardViewModel searchItem = DiscoveryCardViewModel.SearchCard.INSTANCE;

    @Before
    public void setUp() {
        adapter = new DiscoveryAdapter(searchItemRenderer,
                                       singleSelectionContentCardRenderer,
                                       multipleContentSelectionCardRenderer,
                                       emptyCardRenderer,
                                       searchListener);
    }

    @Test
    public void rendersCorrectViewTypes() {
        adapter.onNext(asList(searchItem, singleContentSelectionCard, multipleContentSelectionCard));

        assertThat(adapter.getBasicItemViewType(0)).isEqualTo(DiscoveryCard.Kind.SEARCH_ITEM.ordinal());
        assertThat(adapter.getBasicItemViewType(1)).isEqualTo(DiscoveryCard.Kind.SINGLE_CONTENT_SELECTION_CARD.ordinal());
        assertThat(adapter.getBasicItemViewType(2)).isEqualTo(DiscoveryCard.Kind.MULTIPLE_CONTENT_SELECTION_CARD.ordinal());
    }
}

File: app/src/test/java/com/soundcloud/android/discovery/DiscoveryAdapterTest.kt
package com.soundcloud.android.discovery

import com.soundcloud.android.discovery.DiscoveryAdapter.Kind
import com.soundcloud.android.search.SearchItemRenderer
import com.soundcloud.android.testsupport.AndroidUnitTest
import org.assertj.core.api.Java6Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import java.util.Arrays.asList

class DiscoveryAdapterTest : AndroidUnitTest() {

    @Mock internal lateinit var searchItemRenderer: SearchItemRenderer<DiscoveryCardViewModel>
    @Mock internal lateinit var singleSelectionContentCardRenderer: SingleSelectionContentCardRenderer
    @Mock internal lateinit var multipleContentSelectionCardRenderer: MultipleContentSelectionCardRenderer
    @Mock internal lateinit var emptyCardRenderer: EmptyCardRenderer
    @Mock internal lateinit var searchListener: SearchItemRenderer.SearchListener

    private val multipleContentSelectionCard = DiscoveryFixtures.multiContentSelectionCardViewModel()
    private val singleContentSelectionCard = DiscoveryFixtures.singleContentSelectionCardViewModel()

    private lateinit var adapter: DiscoveryAdapter
    private val searchItem = DiscoveryCardViewModel.SearchCard

    @Before
    fun setUp() {
        adapter = DiscoveryAdapter(searchItemRenderer,
                                   singleSelectionContentCardRenderer,
                                   multipleContentSelectionCardRenderer,
                                   emptyCardRenderer,
                                   searchListener)
    }

    @Test
    fun rendersCorrectViewTypes() {
        adapter.onNext(asList(searchItem, singleContentSelectionCard, multipleContentSelectionCard))

        assertThat(adapter.getBasicItemViewType(0)).isEqualTo(Kind.SEARCH_ITEM.ordinal)
        assertThat(adapter.getBasicItemViewType(1)).isEqualTo(Kind.SINGLE_CONTENT_SELECTION_CARD.ordinal)
        assertThat(adapter.getBasicItemViewType(2)).isEqualTo(Kind.MULTIPLE_CONTENT_SELECTION_CARD.ordinal)
    }
}
